refactor(indexer/base): move to cosmossdk.io/schema

aaronc opened the pull request:
# Description

* Refactor #20629 into `cosmossdk.io/schema` so that this is a more generic logical schema module also suitable for genesis. Schema types were put in the root package and `Listener` was put in the `appdata` sub-package.
* Rename `DecodableModule` to `HasModuleCodec`.
* Rename `ModuleDecoder` to `ModuleCodec`

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review. -->

---

## Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title, you can find examples of the prefixes below:
    <!-- * `feat`: A new feature
    * `fix`: A bug fix
    * `docs`: Documentation only changes
    * `style`: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
    * `refactor`: A code change that neither fixes a bug nor adds a feature
    * `perf`: A code change that improves performance
    * `test`: Adding missing tests or correcting existing tests
    * `build`: Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
    * `ci`: Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
    * `chore`: Other changes that don't modify src or test files
    * `revert`: Reverts a previous commit -->
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] provided a link to the relevant issue or specification
* [ ] reviewed "Files changed" and left comments if necessary
* [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
* [ ] added a changelog entry to `CHANGELOG.md`
* [ ] updated the relevant documentation or specification, including comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] confirmed all CI checks have passed

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

Please see [Pull Request Reviewer section in the contributing guide](../CONTRIBUTING.md#reviewer) for more information on how to review a pull request.

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic, API design and naming, documentation is accurate, tests and test coverage


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Documentation**
  - Added new documentation for the state schema framework in the `cosmossdk.io/schema` base module.
  - Introduced documentation for the `listener` package detailing the streaming framework for blockchain event and state data.
  - Added a README for the `schema/appdata` package explaining the `Listener` callback order for data streaming.

- **Improvement**
  - Renamed parameters and updated types in the `Listener` struct methods to align with the `schema` package's types.
  
- **New Features**
  - Introduced the `HasModuleCodec` interface for logical decoding and encoding in the `schema` package.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

aaronc commented on schema/decoder.go:
```diff
@@ -0,0 +1,35 @@
+	// If it is nil, the module doesn't support state decoding directly.
+	KVDecoder KVDecoder
+
+	// ApplyUpdate is a function that applies an ObjectUpdate to the module's state for the given context.
+	// If it is nil, the module doesn't support applying logical updates. If this function is provided
+	// then it can be used as a genesis import path.
+	ApplyUpdate ApplyUpdate
+}
+
+// KVDecoder is a function that decodes a key-value pair into an ObjectUpdate.
+// If the KV-pair doesn't represent an object update, the function should return false
+// as the second return value. Error should only be non-nil when the decoder expected
+// to parse a valid update and was unable to.
+type KVDecoder = func(key, value []byte) (ObjectUpdate, bool, error)
+
+// ApplyUpdate is a function that applies an ObjectUpdate to the module's state for the given context.
+type ApplyUpdate = func(context.Context, ObjectUpdate) error
```
The `ApplyUpdate` functionality is the key new thing to review here

aaronc commented on schema/listener/listener.go:
```diff
@@ -1,7 +1,9 @@
-package indexerbase
+package listener
```
In the spirit of minimizing the need for extra go modules I've made this a sub-package just called `listener`. Does this seem okay within the context of `schema`? (meaning this is basically `schema/listener` - it's a little weird but maybe okay)

aaronc replied:
What if we called this package `blockdata` and the main type would be `blockdata.Listener`? Seems a bit more logical to have `schema/blockdata.Listener` than `schema/listener.Listener`

kocubinski commented on schema/decoder.go:
```diff
@@ -0,0 +1,35 @@
+	// ApplyUpdate is a function that applies an ObjectUpdate to the module's state for the given context.
+	// If it is nil, the module doesn't support applying logical updates. If this function is provided
+	// then it can be used as a genesis import path.
+	ApplyUpdate ApplyUpdate
```
Apply update provides setting an arbitrary key to an arbitrary value?

aaronc replied:
Basically

kocubinski replied:
So as a module developer, or a streaming developer, if I'm bootstrapping streaming then I would want to somehow wire an export of module state at height _h_ to a a sequence of `ApplyUpdate` executions?

aaronc replied:
What do you mean by streaming developer? If you're bootstrapping a module for testing you could stream a bunch of updates via `ApplyUpdate`. But this wouldn't be for bootstrapping indexers. Is that what you mean?

aaronc replied:
I've dropped this change from this PR to move this along quicker

aaronc commented:
I've updated this PR to just involve moving code and removed the proposed `ApplyUpdate` API. we can get to that in a separate PR. Hopefully this makes this PR simpler to review and move forward.

testinginprod reviewed (approved):
LGTM.
